src/db/db.c

```diff
@@ -452,6 +461,113 @@ dbBackupStop(Db *this)
     FUNCTION_LOG_RETURN_STRUCT(result);
 }
 
+/**********************************************************************************************************************************/
+#define DB_SYNC_TOTAL                                               16
+#define DB_SYNC_SIZE                                                32
+#define DB_SYNC_DATA                                                "elIlg6P7i1yQYtOfsfA4tNSn2io6QK18MzJwzjkBL2QpjuI"
+
+void
+dbSyncCheckHelper(PgClient *const pgClient, const String *const path)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(PG_CLIENT, pgClient);
+        FUNCTION_LOG_PARAM(STRING, path);
+    FUNCTION_LOG_END();
+
+    ASSERT(pgClient != NULL);
+    ASSERT(path != NULL);
+    ASSERT(sizeof(DB_SYNC_DATA) == DB_SYNC_TOTAL + DB_SYNC_SIZE);
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Build data
+        const char *const file = strZ(strNewFmt("%s/pgbackrest.sync.tmp", strZ(path)));
+        const char *fileExpected[DB_SYNC_TOTAL];
+        const String *fileCmd[DB_SYNC_TOTAL];
+
+        for (unsigned int fileIdx = 0; fileIdx < DB_SYNC_TOTAL; fileIdx++)
+        {
+            fileExpected[fileIdx] = strZ(strNewN(DB_SYNC_DATA + fileIdx, DB_SYNC_SIZE));
+            fileCmd[fileIdx] = strNewFmt("copy (select '%s') to '%s'", fileExpected[fileIdx], file);
+
+            ASSERT(fileIdx == 0 || strncmp(fileExpected[fileIdx], fileExpected[fileIdx - 1], DB_SYNC_SIZE) != 0);
+        }
+
+        // Write data and collect results
+        struct stat fileStat[DB_SYNC_TOTAL];
+        char fileActual[DB_SYNC_TOTAL][DB_SYNC_SIZE];
+        ssize_t fileActualSize[DB_SYNC_TOTAL];
+
+        for (unsigned int fileIdx = 0; fileIdx < DB_SYNC_TOTAL; fileIdx++)
+        {
+            // Write data into the file using a copy on the pg server
+            CHECK(pgClientQuery(pgClient, fileCmd[fileIdx]) == NULL);
+
+            // Stat the file
+            THROW_ON_SYS_ERROR_FMT(stat(file, &fileStat[fileIdx]) == -1, FileReadError, "unable to stat '%s'", file);
+
+            // Read the file
+            int fd = open(file, O_RDONLY, 0);
+            THROW_ON_SYS_ERROR_FMT(fd == -1, FileOpenError, "unable to open '%s'", file);
+
+            fileActualSize[fileIdx] = read(fd, fileActual[fileIdx], DB_SYNC_SIZE);
```

Comment:
Reading exactly the size of the data we intended to write, without whatever end of line marker COPY produced, misses the possibility that there could be added junk at the end of the file.